samples/monte-carlo-pi/monte-carlo-pi.cpp

```diff
@@ -0,0 +1,248 @@
+
+    // If work item 0 in work group sum local values
+    if (local_id == 0) {
+      int sum = 0;
+      for (size_t i = 0; i < local_dim; i++) {
+        if (m_local_results_ptr[i] == 1) {
+          ++sum;
+        }
+      }
+      // Store the sum in global memory
+      m_results_ptr[group_id] = sum;
+    }
+  }
+
+ private:
+  read_const_accessor<sycl::cl_float2> m_points_ptr;
+  write_global_accessor<sycl::cl_int> m_results_ptr;
+  read_write_local_accessor<sycl::cl_int> m_local_results_ptr;
+};
+
+/* A helper to define a "perfect" work-group size dependant on selected device
+ * and kernel maximum allowance */
+size_t get_best_work_group_size(size_t work_group_size,
+                                const sycl::device& device,
+                                const sycl::kernel& kernel) {
+  if (device.is_host()) {
+    /* Maximum Work-Group Size on selected device.
+     * (See: sycl-1.2.1.pdf: p.62, table 4.20) */
+    const size_t max_device_work_group_size =
+        device.get_info<cl::sycl::info::device::max_work_group_size>();
+    /* Check if the desired work-group size will be allowed on the host device
+     * and query the maximum possible size on that device in case the desired
+     * one is more than the allowed */
+    if (work_group_size > max_device_work_group_size) {
+      std::cout << "Maximum work-group size for device "
+                << device.get_info<cl::sycl::info::device::name>() << ": "
+                << max_device_work_group_size << "\n";
+      return max_device_work_group_size;
+    }
+    return work_group_size;
+  } else {
+    /* Maximum Work-Group Size for given kernel on selected device.
+     * (See: sycl-1.2.1.pdf: p.180, table 4.85) */
+    const size_t max_kernel_work_group_size = kernel.get_work_group_info<
+        sycl::info::kernel_work_group::work_group_size>(device);
+    /* Verify if the kernel can be executed with our desired work-group size,
+     * and if it can't use the maximum allowed kernel work-group size for the
+     * selected device.
+     */
+    if (work_group_size > max_kernel_work_group_size) {
+      std::cout << "Maximum work-group size for "
+                << typeid(monte_carlo_pi_kernel).name() << " on device "
+                << device.get_info<sycl::info::device::name>() << ": "
+                << max_kernel_work_group_size << "\n";
+      return max_kernel_work_group_size;
+    }
+    // Otherwise, the work-size will stay the originally desired one
+    return work_group_size;
+  }
+}
+
+int main() {
+  constexpr size_t iterations = 1 << 20;
+  size_t work_group_size = 1 << 10;
+
+  // Container for the sum calculated per each work-group.
+  std::vector<sycl::cl_int> results;
+
+  // Generate random points on the host - one point for each thread
+  std::vector<sycl::float2> points(iterations);
+  // Fill up with (pseudo) random values in the range: [0, 1]
+  std::generate(points.begin(), points.end(), []() {
```

Comment:
Since this is a lambda, I think it is better to create the random device and distribution etc. outside of the generate call and have them captured by reference inside the lambda.

Reply:
Definitely better! My initial idea was to have it defined `static` inside so it is slightly more compact, I mean same lines of code but seem more as a part of the lambda but I see how this was not a really good idea. 

Actually, if I have enclosed the whole random generation operation in a scope `{}` would've still done it for me but I guess I didn't think about it as much at the time. 

Anyway, I will just do as suggested. No need to debate on style or this will never get merged lol.

Reply:
Yeah the way you have it now is perfect!